nxcomp/Log.cpp

```diff
@@ -0,0 +1,126 @@
+/*                                                                        */
+/* Redistribution and use of the present software is allowed according    */
+/* to terms specified in the file LICENSE.nxcomp which comes in the       */
+/* source distribution.                                                   */
+/*                                                                        */
+/* All rights reserved.                                                   */
+/*                                                                        */
+/* NOTE: This software has received contributions from various other      */
+/* contributors, only the core maintainers and supporters are listed as   */
+/* copyright holders. Please contact us, if you feel you should be listed */
+/* as copyright holder, as well.                                          */
+/*                                                                        */
+/**************************************************************************/
+
+
+#include <stdio.h>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <iomanip>
+
+#include "Log.h"
+
+NXLog nx_log;
+
+
+bool NXLog::will_log() const
+{
+	std::map<std::string, NXLogLevel>::const_iterator item = per_file_levels_.find(current_file());
+
+	if ( item != per_file_levels_.end() )
+	{
+		return current_level() >= item->second;
+	}
+	else
+	{
+		return current_level() >= level();
+	}
+}
+
+
+std::string NXLog::stamp_to_string(const NXLogStamp& stamp) const
+{
+	std::ostringstream oss;
+
+	static const char* level_names[] = {
+		"DEBUG",
+		"INFO ",
+		"WARN ",
+		"ERROR",
+		"FATAL"
+	};
+
+	if ( log_level() )
+		oss << ((stamp.level() >=0 && stamp.level() < NXLOG_LEVEL_COUNT ) ? level_names[stamp.level()] : "???") << " ";
+
+	if ( log_time() )
+	{
+		struct timeval timestamp = stamp.timestamp();
+		struct tm timeinfo;
+
+		localtime_r(&timestamp.tv_sec, &timeinfo);
+
+		if ( log_unix_time() )
+		{
+			oss << timestamp.tv_sec;
+		}
+		else
+		{
+			#if __cplusplus >= 201103L 
+			oss << " " << std::put_time(&timeinfo, "%Y/%m/%d %H:%M:%S");
+			#else
+			oss << timestamp.tv_sec;
+			#endif
+		}
+
+		oss << "." << std::setw(3) << std::setfill('0') << (int)(timestamp.tv_usec / 1000) << " ";
+	}
+
+	if ( log_location() )
+		oss << stamp.file() << "/" << stamp.function() << ":" << stamp.line()  << " ";
+
+	if ( log_thread_id() )
+	{
+		if ( thread_name().empty() )
+			oss << pthread_self() << " ";
+		else
+			oss << "[" << thread_name() << "] ";
+	}
+
+	return oss.str();
+}
+
+NXLog& operator<< (NXLog& out, const NXLogStamp& value)
+{
+	out.current_level( value.level() );
+	out.current_file( value.file() );
+
+	if (  out.will_log() )
+	{
+		if ( out.synchronized() )
+		{
+			out.flush(); // Flush the previous line
+			NXLog::per_thread_data *pdt = out.get_data();
+			(*pdt->buffer) << out.stamp_to_string(value);
+		}
+		else
+		{
+			*(out.stream()) << out.stamp_to_string(value) << " ";
```

Comment:
But that's not going to work. Streams work in chains, and there's no way to tell where it ends. I mean that this:

    stream << "User " << username << " logged in" << std::endl;

Is exactly the same thing as:

    stream << "User ";
    stream << username;
    stream << " logged in";
    stream << std::endl;

Internally an operator<< returns a stream, so what you're doing is:

    stream.write("User ").write(username).write(" logged in").write(std::endl);

As far as anything in the logging code can tell there are 4 write calls. But we want to somehow group multiple write calls into logical log messages, otherwise output from multiple threads at once will be a complete mess. So you need to some kind of terminating condition.

This, by the way, is why the opposite of std::endl makes no sense -- that's how everything but \n and std::endl works already.

So for terminating conditions we have: std::endl, "\n" in whatever is written, too much in buffer, and the write of a NXLogStamp. The later could add a missing "\n" but it would be dangerous, Suppose:

    nx_log << nxstamp(NXINFO) << "I am about to divide by " << n;
    foo = 5/n;
    nx_log << nxstamp(NXINFO) << "Done.\n";


Normally the output would be just fine. But if n is 0, you get a crash, and the first message remains in the buffer and never gets sent to the output. So you don't want to fix the first log message automatically and hide the problem. Rather you want to make it look obviously wrong so that it gets fixed.